daemons/execd/pacemaker-execd.c

```diff
@@ -400,6 +404,66 @@ handle_shutdown_nack(void)
     crm_debug("Ignoring unexpected shutdown nack");
 }
 
+#ifdef PCMK__COMPILE_REMOTE
+static int
+rm_files(const char *pathname, const struct stat *sbuf, int type, struct FTW *ftwb)
+{
+    /* Don't delete CRM_REMOTE_SCHEMA_DIRECTORY . */
+    if (pcmk__str_eq(pathname, CRM_REMOTE_SCHEMA_DIRECTORY, pcmk__str_none)) {
+        return 0;
+    }
+
+    if (remove(pathname) != 0) {
+        int rc = errno;
+        crm_err("Could not remove %s: %s", pathname, pcmk_rc_str(rc));
+        return -1;
+    }
+
+    return 0;
+}
+
+static void
+clean_up_extra_schema_files(void)
+{
+    struct stat sb;
+    int rc;
+
+    rc = stat(CRM_REMOTE_SCHEMA_DIRECTORY, &sb);
+
+    if (rc == -1) {
+        if (errno == ENOENT) {
+            /* If the directory doesn't exist, try to make it first. */
+            if (mkdir(CRM_REMOTE_SCHEMA_DIRECTORY, 0755) != 0) {
+                rc = errno;
+                crm_err("Could not create directory for schemas: %s",
+                        pcmk_rc_str(rc));
+                return;
+            }
+        } else {
+            rc = errno;
+            crm_err("Could not create directory for schemas: %s",
+                    pcmk_rc_str(rc));
+            return;
+        }
+    } else if (!S_ISDIR(sb.st_mode)) {
+        /* If something exists with the same name that's not a directory, that's
+         * an error.
+         */
+        crm_err("%s already exists but is not a directory", CRM_REMOTE_SCHEMA_DIRECTORY);
+        return;
+    } else {
+        /* It's a directory - clear it out so we can download potentially new
+         * schema files.
+         */
+        rc = nftw(CRM_REMOTE_SCHEMA_DIRECTORY, rm_files, 10, FTW_DEPTH|FTW_MOUNT|FTW_PHYS);
+
+        if (rc != 0) {
+            crm_err("Could not remove %s: %s", CRM_REMOTE_SCHEMA_DIRECTORY, pcmk_rc_str(rc));
+        }
+    }
+}
+#endif
+
```

Comment:
There's a couple things in here that could potentially be moved into library functions:

* Removing a directory tree
* Creating a directory if it doesn't exist
* Testing if a directory exists or not